aws/rust-runtime/aws-config/src/profile/token.rs

```diff
@@ -0,0 +1,248 @@
+///
+/// Generally, this provider is constructed via the default provider chain. However,
+/// it can also be manually constructed with the builder:
+/// ```rust,no_run
+/// use aws_config::profile::ProfileFileTokenProvider;
+/// let provider = ProfileFileTokenProvider::builder().build();
+/// ```
+/// _Note: this provider, when called, will load and parse the `~/.aws/config` file
+/// only once. Parsed file contents will be cached indefinitely._
+///
+/// This provider requires a profile with a `sso_session` configured. For example,
+/// ```ini
+/// [default]
+/// sso_session = example
+/// region = us-west-2
+///
+/// [sso-session example]
+/// sso_start_url = https://example.awsapps.com/start
+/// sso_region = us-west-2
+/// ```
+#[derive(Debug)]
+pub struct ProfileFileTokenProvider {
+    sdk_config: SdkConfig,
+    provider_config: ProviderConfig,
+    inner_provider: ErrorTakingOnceCell<SsoTokenProvider, TokenError>,
+}
+
+impl ProfileFileTokenProvider {
+    /// Builder for this token provider.
+    pub fn builder() -> Builder {
+        Builder::default()
+    }
+
+    async fn load_token(&self) -> TokenResult {
+        let inner_provider = self
+            .inner_provider
+            .get_or_init(
+                {
+                    let sdk_config = self.sdk_config.clone();
+                    let provider_config = self.provider_config.clone();
+                    move || async move {
+                        let profile_set = load_profile_set(&provider_config).await?;
+                        create_token_provider(&sdk_config, &provider_config, profile_set)
+                    }
+                },
+                TokenError::unhandled(
+                    "profile file token provider initialization error already taken",
+                ),
+            )
+            .await?;
+
+        inner_provider.provide_token().await
+    }
+}
+
+impl ProvideToken for ProfileFileTokenProvider {
+    fn provide_token<'a>(&'a self) -> future::ProvideToken<'a>
+    where
+        Self: 'a,
+    {
+        future::ProvideToken::new(self.load_token())
+    }
+}
+
+/// Builder for [`ProfileFileTokenProvider`].
+#[derive(Debug, Default)]
+pub struct Builder {
+    provider_config: Option<ProviderConfig>,
+    profile_override: Option<String>,
+    profile_files: Option<ProfileFiles>,
+}
+
+impl Builder {
+    /// Override the configuration for the [`ProfileFileTokenProvider`]
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # async fn test() {
+    /// use aws_config::profile::ProfileFileTokenProvider;
+    /// use aws_config::provider_config::ProviderConfig;
+    /// let provider = ProfileFileTokenProvider::builder()
+    ///     .configure(&ProviderConfig::with_default_region().await)
+    ///     .build();
+    /// # }
+    /// ```
+    pub fn configure(mut self, provider_config: &ProviderConfig) -> Self {
+        self.provider_config = Some(provider_config.clone());
+        self
+    }
+
```

Comment:
I missed this in first reading. We shouldn't expose new functions that expose `ProviderConfig` publicly. Prefer `SdkConfig` or have this method be private.

Reply:
Fixed in 89a2ea8895f32066e1ad803ec634683c307b0c84